File: CTDash/managers/transport_manager.py
# [BUILD FIRST]
# TransportManager — handles all patient movement delays.
# Runs the Transport state machine for each active patient.
# Generates random delay values within config ranges on patient spawn.

import random
from config import TRANSPORT_ARRIVAL_DELAY, TRANSPORT_HOLD_WAIT, TRANSPORT_LEAVING_DELAY
from classes.patient import PatientState
from classes.transport import TransportState


class TransportManager:
    def __init__(self):
        self._active = {}    # patient_id -> Patient (those currently in transport)

    def assign_transport(self, patient):
        """Called when a patient is ordered. Sets random delay values and activates."""
        t = patient.transport
        t.arrival_delay  = random.randint(*TRANSPORT_ARRIVAL_DELAY)
        t.hold_wait      = random.randint(*TRANSPORT_HOLD_WAIT)
        t.leaving_delay  = random.randint(*TRANSPORT_LEAVING_DELAY)
        t.timer          = t.arrival_delay
        t.state          = TransportState.WAITING_ASSIGNMENT
        patient.state    = PatientState.IN_TRANSPORT
        self._active[patient.patient_id] = patient

    def tick(self):
        """Advance transport state machines for all active patients."""
        delivered = []

        for pid, patient in self._active.items():
            t = patient.transport
            t.timer -= 1

            if t.timer > 0:
                continue

            # Timer expired — advance state.
            if t.state == TransportState.WAITING_ASSIGNMENT:
                # Transporter acknowledged; now heading to patient.
                t.state = TransportState.ACKNOWLEDGED
                t.timer = t.hold_wait

            elif t.state == TransportState.ACKNOWLEDGED:
                # Transporter at patient, loading and moving.
                t.state = TransportState.ON_WAY
                t.timer = 0

            elif t.state == TransportState.ON_WAY:
                t.state = TransportState.DELIVERED
                # LEAVING patients are departing; all others arrive at holding bay.
                if patient.state == PatientState.LEAVING:
                    patient.state = PatientState.COMPLETED
                else:
                    patient.state = PatientState.IN_HOLDING
                delivered.append(pid)

        for pid in delivered:
            del self._active[pid]

    def begin_leaving(self, patient):
        """Called after scan completes. Patient waits leaving_delay then departs."""
        t = patient.transport
        # Jump straight to ON_WAY with leaving_delay so only one phase runs.
        t.state = TransportState.ON_WAY
        t.timer = t.leaving_delay
        patient.state = PatientState.LEAVING
        self._active[patient.patient_id] = patient

File: CTDash_Unity/Assets/Scripts/Managers/TransportManager.cs
// TransportManager — handles all patient movement delays.
// Runs the Transport state machine for each active patient.

using System.Collections.Generic;

namespace CTDash
{
    public class TransportManager
    {
        private readonly Dictionary<string, Patient> _active = new Dictionary<string, Patient>();

        public void AssignTransport(Patient patient)
        {
            var t         = patient.Transport;
            t.State       = TransportState.WaitingAssignment;
            t.Timer       = t.ArrivalDelay;
            patient.State = PatientState.InTransport;
            _active[patient.PatientId] = patient;
        }

        public void Tick()
        {
            var delivered = new List<string>();

            foreach (var kvp in _active)
            {
                var patient = kvp.Value;
                var t       = patient.Transport;
                t.Timer    -= 1;

                if (t.Timer > 0) continue;

                switch (t.State)
                {
                    case TransportState.WaitingAssignment:
                        t.State = TransportState.Acknowledged;
                        t.Timer = t.HoldWait;
                        break;

                    case TransportState.Acknowledged:
                        t.State = TransportState.OnWay;
                        t.Timer = 0;
                        break;

                    case TransportState.OnWay:
                        t.State = TransportState.Delivered;
                        if (patient.State == PatientState.Leaving)
                            patient.State = PatientState.Completed;
                        else
                            patient.State = PatientState.InHolding;
                        delivered.Add(kvp.Key);
                        break;
                }
            }

            foreach (var pid in delivered)
                _active.Remove(pid);
        }

        public void BeginLeaving(Patient patient)
        {
            var t         = patient.Transport;
            t.State       = TransportState.OnWay;
            t.Timer       = t.LeavingDelay;
            patient.State = PatientState.Leaving;
            _active[patient.PatientId] = patient;
        }
    }
}
